src/v/cluster/partition.h

```diff
@@ -93,20 +93,38 @@ class partition {
     }
 
     ss::future<result<model::offset, std::error_code>>
-    sync_effective_start(model::timeout_clock::duration timeout) {
+    sync_kafka_start_offset_override(model::timeout_clock::duration timeout) {
         if (_log_eviction_stm) {
-            co_return co_await _log_eviction_stm->sync_effective_start(timeout);
+            auto offset_res = co_await _log_eviction_stm
+                                ->sync_kafka_start_offset_override(timeout);
+            if (offset_res.has_failure()) {
+                co_return offset_res.as_failure();
+            }
+            auto offset_translator_state = get_offset_translator_state();
+            if (_raft->start_offset() < offset_res.value()) {
+                auto start_kafka_offset
+                  = offset_translator_state->from_log_offset(
+                    offset_res.value());
+                co_return start_kafka_offset;
+            }
+            // If a start override is no longer in the offset translator state,
+            // it may have been uploaded andpersisted in the manifest.
         }
-        co_return raft_start_offset();
-    }
-
-    model::offset raft_start_offset() const {
-        if (_log_eviction_stm) {
-            return _log_eviction_stm->effective_start_offset();
+        if (_archival_meta_stm) {
+            if (!co_await _archival_meta_stm->sync(timeout)) {
+                co_return errc::not_leader;
```

Comment:
I can see how this is sort of practical error code, but a failure of the STM sync isn't literally telling us we're not the leader, right?  It could also mean that sync was just too slow for whatever reason.

I'm wondering if there might be a situation where returning `not_leader` on a very slow system could be harmful, because it prompts a kafka client to issue more requests trying to refresh metadata.